rfd/0049-sso-configure-command.md

```diff
@@ -0,0 +1,146 @@
+| `oidc`   |            | Ask for `--preset`.                                       | Default to `--preset=generic`.     |
+| `oidc`   | `generic`  | Generic OIDC template.                                    |                                    |
+| `oidc`   | `okta`     | Okta OIDC template. Fill in details from `.well-known`.   |                                    |
+| `oidc`   | `google`   | Google OIDC template. Fill in details from `.well-known`. |                                    |
+| `saml`   |            | Ask for `--preset`.                                       | Default to `--preset=generic`.     |
+| `saml`   | `generic`  | Generic SAML template.                                    |                                    |
+| `saml`   | `okta`     | Okta SAML template.                                       |                                    |
+| `saml`   | `google`   | Google SAML template.                                     |                                    |
+| `saml`   | `adfs`     | ADFS SAML template.                                       |                                    |
+
+Individual templates can accept additional non-standard flags. For example `--entity-descriptor=FILE` flag used by SAML
+templates will read Entity Descriptor XML from the provided file and properly embed it in the resulting YAML file. The
+parameters *may* be subject to validation (e.g. XML structure being correct), but by default incorrect parameters should
+only cause warnings to be emitted. In general, we cannot guarantee correctness of generated config; it is expected that
+the user will review the config file and test it for correctness. This is especially true for things
+like `attributes_to_roles` mappings or the provider side of the configuration. This tool is meant to make the
+configuration easier, not replace the appropriate documentation.
+
+Detailed help message can be access with:
+
+```bash
+$ tctl sso configure --help
+...
+```
+
+## Interactive and batch modes
+
+We support two modes: interactive and batch (non-interactive). In interactive mode we may provide the user with
+choices (e.g. preset) or ask them to provide details (secrets, identifiers etc.) In batch mode we never ask but use
+defaults (when possible) or insert comments asking for the data to be filled manually afterwards. Some
+
+Interactive mode is used when we detect TTY or `--interactive` flag is provided, batch mode is used otherwise.
+
+## Proxy URL discovery
+
+The command will make an effort to discover the correct proxy URL. Failing that, an appropriate default along with
+comment will be put into the configuration file.
+
+## Implementation details
+
+Each combination of `--kind` and `--provider` will be mapped to single template based
+on [text/template](https://pkg.go.dev/text/template) package. The extension points will be handled by calls to functions
```

Comment:
Sure, this makes things simpler if less flexible (we cannot output ill-formed YAML with those methods, so there is no way of adding comments and similar stuff). But this may be a good thing after all: the implementation will be simpler and more testable.